Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+    unitForce_G[1] = -Vec3(ppe.terminationTangent_G);
+}
+
+} // namespace
+
+void CableSpan::Impl::applyBodyForces(
+    const State& s,
+    Real tension,
+    Vector_<SpatialVec>& bodyForcesInG) const
+{
+    realizePosition(s);
+
+    if (tension < 0.) {
+        return;
+    }
+
+    SpatialVec unitForce_G;
+
+    // Force applied at cable origin point.
+    {
+        calcUnitForceAtCableOrigin(*this, s, unitForce_G);
+        getOriginBody().applyBodyForce(s, unitForce_G * tension, bodyForcesInG);
+    }
+
+    // Forces applied to each obstacle body.
+    for (const CurveSegment& curve : m_CurveSegments) {
+        if (!curve.isInContactWithSurface(s)) {
+            continue;
+        }
+
+        calcUnitForceExertedByCurve(curve, s, unitForce_G);
+        curve.getMobilizedBody().applyBodyForce(
+            s,
+            unitForce_G * tension,
+            bodyForcesInG);
+    }
+
+    // Force applied at cable termination point.
+    {
+        calcUnitForceAtCableTermination(*this, s, unitForce_G);
+        getTerminationBody().applyBodyForce(
+            s,
+            unitForce_G * tension,
+            bodyForcesInG);
+    }
+}
+
+Real CableSpan::Impl::calcCablePower(const State& state, Real tension) const
+{
+    realizePosition(state);
+
+    if (tension < 0.) {
+        return 0.;
+    }
+
+    SpatialVec unitForce_G;
+
+    Real unitPower = 0.;
+
+    {
+        calcUnitForceAtCableOrigin(*this, state, unitForce_G);
+        SpatialVec v = getOriginBody().getBodyVelocity(state);
+        unitPower += ~unitForce_G * v;
+    }
+
+    for (const CurveSegment& curve : m_CurveSegments) {
+        if (!curve.isInContactWithSurface(state)) {
+            continue;
+        }
+
+        calcUnitForceExertedByCurve(curve, state, unitForce_G);
+        SpatialVec v = curve.getMobilizedBody().getBodyVelocity(state);
+        unitPower += ~unitForce_G * v;
+    }
+
+    {
+        calcUnitForceAtCableTermination(*this, state, unitForce_G);
+        SpatialVec v = getTerminationBody().getBodyVelocity(state);
+        unitPower += ~unitForce_G * v;
+    }
+
+    return unitPower * tension;
+}
+
+//==============================================================================
+//                             CABLE VISUALIZATION
+//==============================================================================
+
+namespace
+{
+
+// Helper for drawing a pretty curve segment.
+void calcCurveDecorativeGeometryAndAppend(
+    const CurveSegment& curve,
+    const State& s,
+    Array_<DecorativeGeometry>& decorations)
+{
+    static constexpr int LINE_THICKNESS    = 3;
+    static constexpr Real INACTIVE_OPACITY = 0.25;
+
+    const bool isInContactWithSurface = curve.isInContactWithSurface(s);
+    const CurveSegment::PosEntry& ppe = curve.getPosInfo(s);
+
+    // Draw the surface (TODO since we do not own it, should it be done here at
+    // all?).
+    {
+        DecorativeGeometry geo = curve.getDecoration(); // TODO clone it?
```

Comment:
Or at least handle it as a reference - this class looks sliced?

Reply:
I changed the `CableSpan` to take the `ContactGeometry` by shared reference, when adding an obstacle to the path. This way a single `ContactGeometry` can be reused as obstacle in multiple paths. I therefore removed rendering the surface geometry from the `CableSpan::calcDecorativeGeometry`.